src/pages/o/[orgId]/surveys/[surveyId]/index.tsx

```diff
@@ -1,32 +1,346 @@
+  ZetkinSurveyTextQuestionElement,
+} from 'utils/types/zetkin';
 
 const scaffoldOptions = {
   allowNonOfficials: true,
   authLevelRequired: 1,
 };
 
+function parseRequest(
+  req: IncomingMessage
+): Promise<NodeJS.Dict<string | string[]>> {
+  return new Promise((resolve) => {
+    let body = '';
+    req.on('data', (chunk: Buffer) => {
+      body += chunk.toString();
+    });
+    req.on('end', () => {
+      resolve(parse(body));
+    });
+  });
+}
+
 export const getServerSideProps = scaffold(async (ctx) => {
+  const { req, res } = ctx;
   const { surveyId, orgId } = ctx.params!;
+  let status: FormStatus = 'editing';
+
+  const apiClient = new BackendApiClient(req.headers);
+  let survey: ZetkinSurveyExtended;
+  try {
+    survey = await apiClient.get<ZetkinSurveyExtended>(
+      `/api/orgs/${orgId}/surveys/${surveyId}`
+    );
+  } catch (e) {
+    return { notFound: true };
+  }
+
+  if (req.method === 'POST') {
+    const form = await parseRequest(req);
+    const responses: Record<
+      string,
+      {
+        options?: number[];
+        question_id: number;
+        response?: string;
+      }
+    > = {};
+
+    for (const name in form) {
+      const isSignature = name.startsWith('sig');
+      const isPrivacy = name.startsWith('privacy');
+      const isMetadata = isSignature || isPrivacy;
+      if (isMetadata) {
+        continue;
+      }
+
+      const fields = name.split('.');
+      const questionId = fields[0];
+      const questionType = fields[1];
+
+      if (typeof responses[questionId] === 'undefined') {
+        responses[questionId] = {
+          question_id: parseInt(fields[0]),
+        };
+      }
+
+      if (questionType == 'options') {
+        if (Array.isArray(form[name])) {
+          responses[questionId].options = (form[name] as string[]).map((o) =>
+            parseInt(o)
+          );
+        } else {
+          responses[questionId].options = [parseInt((form[name] as string)!)];
+        }
+      } else if (questionType == 'text') {
+        responses[questionId].response = form[name] as string;
+      }
+    }
+
+    let signature: null | {
+      email: string;
+      first_name: string;
+      last_name: string;
+    } = null;
+    // TODO: handle other signature types
+    if (form.sig == 'email') {
+      signature = {
+        email: form['sig.email'] as string,
+        first_name: form['sig.first_name'] as string,
+        last_name: form['sig.last_name'] as string,
+      };
+    }
+
+    try {
+      await apiClient.post(
+        `/api/orgs/${orgId}/surveys/${surveyId}/submissions`,
+        {
+          responses: Object.values(responses),
+          signature,
+        }
+      );
+    } catch (e) {
+      status = 'error';
+    }
+    if (status !== 'error') {
+      status = 'submitted';
+    }
+
+    if (status === 'submitted') {
+      res.writeHead(302, {
+        Location: `/o/${orgId}/surveys/${surveyId}/submitted`,
+      });
+      res.end();
```

Comment:
The ["Next.js way"](https://nextjs.org/docs/pages/api-reference/functions/get-server-side-props#redirect) to redirect is by returning an object with the `redirect` property (instead of `props`), like this:

```ts
return {
  redirect: {
    destination: `/o/${orgId}/surveys/${surveyId}/submitted`,
    permanent: false,
  },
};
```